enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,290 @@
+---
+title: operator-status
+authors:
+  - "@awgreene"
+reviewers:
+  - "@ecordell"
+approvers:
+  - "@kevinrizza"
+creation-date: 2020-05-19
+last-updated: yyyy-mm-dd
+status: implementable
+see-also:
+  - "N/A"  
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# operator-status
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+Managing the upgrade of an operator is a core feature of the [Operator Lifecycle Manager (OLM) project](https://github.com/operator-framework/operator-lifecycle-manager). There are instances where an operator is performing a long running critical task during which it should not be interrupted by an upgrade initiated by OLM. OLM must provide operators with the means to communicate whether or not the operator is ready to be upgraded.
+
+## Motivation
+
+Today, OLM does not provide operators with the ability to block upgrades as the operator completes a critical process. Consider the following example:
+
+An operator has three versions available, v1, v2, and v3 where v2 has a long-running process that migrates resources in v1 to the new schema in v2. As of today, if operator v1 was installed with automatic updates enabled, and v2 and v3 became available at the same time, OLM would upgrade the operator from v1 to v2 to v3 without allowing v2's migration process to run. 
+
+The scenario above could place the operator in an unrecoverable state. OLM must provide operators that it manages with a method to communicate whether or not the operator is ready to be upgraded. 
+
+### Goals
+
+- Provide opreators with the means to communicate whether or not they should be upgraded.
+- Prevent operators from being upgraded if they communicate that they should not be upgraded.
+- Preserve existing behavior should the operator not communicate its upgradeable status.
+
+### Non-Goals
+
+- Support operator version rollback.
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade my operator until I remove the flag.
+
+### Implementation Details/Notes/Constraints
+
```

Comment:
What do you think about this idea?

The operator can signal it's readiness using the existing condition array in the status field. This does not requires API changes.

e.g.
``` json
  {
    "lastHeartbeatTime": "2020-06-08T13:01:39Z",
    "lastTransitionTime": "2020-06-08T12:59:12Z",
    "message": "Reconcile completed successfully",
    "reason": "ReconcileCompleted",
    "status": "True",
    "type": "Upgradeable"
  }
```

Reply:
The status of the CSV? That seems like it would break the pattern of status ownership on the operator of the object (the OLM operator). Also, what about when that operator is running in an environment where OLM does not exist? Or when OLM migrates to a different version of the API? I would be concerned about a tight coupling here that I think we should try to avoid.

Reply:
This places the upgrade readiness on the collection of CRs owned by the operator in question. I think it's technically possible (although discouraged from what I understand) for an operator not to have a CRD associated with it.

Reply:
No. The CR's status

Reply:
The latest version of this PR is very similar to this suggestion, please review.

Reply:
:+1: 

I also liked the probe CR addition, though our CR is a singleton and we don't need it.

This actually what we implemented in KubeVirt's hyperconverged-cluster-operator (HCO): https://github.com/kubevirt/hyperconverged-cluster-operator/blob/beccfd1fb274e283c86b848a5109d376e1cb3c7f/pkg/controller/hyperconverged/hyperconverged_controller.go#L501